src/main.rs

```diff
@@ -17,53 +17,110 @@ use std::env;
+    letterboxd-sync [--recursive] --pattern=<regex> <list-id> <folder>
 
 Options:
     --pattern=<regex>  The pattern used to extract the movie names.
+    -r --recursive     Search for movies in the given folder recursively.
 ";
 
 #[derive(Debug, Deserialize)]
 struct Args {
     flag_pattern: String,
     arg_list_id: String,
     arg_folder: String,
+    flag_recursive: bool,
 }
 
 /// Returns true if entry is a file, false otherwise or on error.
 fn is_file(entry: &fs::DirEntry) -> bool {
     entry.metadata().ok().map(|m| m.is_file()).unwrap_or(false)
 }
 
-/// Used in to filter_map entries into files. Directories are sorted out.
-fn into_file(entry: io::Result<fs::DirEntry>) -> Option<fs::DirEntry> {
-    entry.ok().and_then(
-        |e| if is_file(&e) { Some(e) } else { None },
-    )
+struct Files {
+    entries: fs::ReadDir,
+    stack: Vec<PathBuf>,
+    recursively: bool,
+}
+
+impl Files {
+    /// Returns the next file in the next dir on the stack if any.
+    fn next_in_dir(&mut self) -> Option<String> {
+        self.stack.pop().and_then(|dir| {
+            match fs::read_dir(&dir) {
+                Ok(new_entries) => self.entries = new_entries,
+                Err(_) => println!("[W] Could not read files in {:?}", dir),
+            }
+            self.next()
+        })
+    }
+
+    /// Return file of possible entry or move on to next.
+    fn handle_next(&mut self, entry: fs::DirEntry) -> Option<String> {
+        let path = entry.path();
+        if is_file(&entry) {
+            match entry.file_name().into_string() {
+                Ok(filename) => Some(filename),
+                Err(filename) => {
+                    println!("[W] Could not retrieve filename of {:?}", filename);
+                    self.next()
+                }
+            }
+        } else if self.recursively && path.is_dir() {
+            self.stack.push(path.to_path_buf());
+            self.next()
+        } else {
+            None
+        }
+    }
+}
+
+impl Iterator for Files {
+    type Item = String;
+
+    fn next(&mut self) -> Option<String> {
+        match self.entries.next() {
+            None => self.next_in_dir(),
+            Some(maybe_entry) => {
+                if let Ok(entry) = maybe_entry {
+                    self.handle_next(entry)
+                } else {
+                    self.next()
+                }
+            }
+        }
+    }
 }
 
 /// List all files in dir.
-fn list_files(
-    path: &str,
-) -> Result<impl Iterator<Item = std::string::String>, Box<std::error::Error>> {
-    let dir = fs::read_dir(path)?;
-    let files = dir.filter_map(into_file).filter_map(|e| {
-        e.file_name().into_string().ok()
-    });
-    Ok(files)
+fn list_files(dir: &str, recursively: bool) -> Result<Vec<String>, Box<std::error::Error>> {
+    let path = Path::new(dir);
+    if !path.is_dir() {
+        return Ok(vec![String::from(dir)]);
+    }
+
+    let entries = fs::read_dir(path)?;
```

Comment:
Ich habe irgendwie das Gefuehl, dass das doppelt ist. Eigentlich wuerde es ja ausrecihen, dir in den Stack zu legen, nicht?

Reply:
Ja, hatte ich auch schon irgendwie überlegt.